lib/mayaUsd/render/MaterialXGenOgsXml/ShaderGenUtil.cpp

```diff
@@ -0,0 +1,383 @@
+
+        auto        destNodeIt = _nodeMap.find(sourceNode.getNamePath());
+        mx::NodePtr destNode;
+        if (destNodeIt != _nodeMap.end()) {
+            destNode = destNodeIt->second;
+        } else {
+            if (!_nodeGraph) {
+                _nodeGraph = _doc->addNodeGraph("NG0");
+            }
+            destNode = cloneNode(sourceNode, *_nodeGraph);
+        }
+
+        auto sourceNodeDef = sourceNode.getNodeDef();
+        if (!sourceNodeDef) {
+            throw mx::Exception("Could not find NodeDef.");
+        }
+
+        const bool isTopological = isTopologicalNodeDef(*sourceNodeDef);
+        for (const auto& defInput : sourceNodeDef->getActiveInputs()) {
+            auto sourceInput = sourceNode.getInput(defInput->getName());
+            if (!sourceInput) {
+                continue;
+            }
+
+            auto connectedNode = sourceInput->getConnectedNode();
+            if (connectedNode) {
+                auto        destConnectedIt = _nodeMap.find(connectedNode->getNamePath());
+                mx::NodePtr destConnectedNode;
+                if (destConnectedIt != _nodeMap.end()) {
+                    destConnectedNode = destConnectedIt->second;
+                } else {
+                    if (!_nodeGraph) {
+                        _nodeGraph = _doc->addNodeGraph("NG0");
+                    }
+                    destConnectedNode = cloneNode(*connectedNode, *_nodeGraph);
+                    nodesToTraverse.push_back(connectedNode);
+                }
+
+                const std::string channelInfo = gatherChannels(*sourceInput);
+                const std::string outputString = gatherOutput(*sourceInput);
+
+                if (sourceNode != *surfaceShader) {
+                    cloneConnection(
+                        *sourceInput, *destNode, destConnectedNode, channelInfo, outputString);
+                } else {
+                    cloneNodeGraphConnection(
+                        *sourceInput, *destNode, destConnectedNode, channelInfo, outputString);
+                }
+            } else if (isTopological) {
+                std::string valueString = sourceInput->getValueString();
+                if (valueString.empty()) {
+                    const auto interfaceInput = sourceInput->getInterfaceInput();
+                    if (interfaceInput) {
+                        valueString = interfaceInput->getValueString();
+                    }
+                }
+                if (!valueString.empty()) {
+                    auto destInput
+                        = destNode->addInput(sourceInput->getName(), sourceInput->getType());
+                    destInput->setValueString(valueString);
+                }
+            }
+        }
+    }
+}
+
+mx::NodePtr TopoNeutralGraph::cloneNode(const mx::Node& node, mx::GraphElement& container)
+{
+    auto destNode
+        = container.addNode(node.getCategory(), "N" + std::to_string(_nodeIndex), node.getType());
```

Comment:
All cloned nodes get incremental names starting at N0. This allows merging graphs that have the same nodes and connectivity, but differs in node names.